### PR TITLE
Added attention model

### PR DESCRIPTION
**Pull request by victor-armegioiu**
Attention based sequence classifier

## Overview
Generic sequence text classifier, based on Luong's multiplicative style attention mechanism. Tested it on the Cornell Subjectivity Dataset (http://www.cs.cornell.edu/people/pabo/movie-review-data/) and achieved 93% accuracy on the test set. Results may vary depending on the the parameters initialization.

## Testing Instructions
The model accepts training/testing data of the form (X_train, y_train) where X_train = list of lists of indexed tokens; y_train = list of lists of binarized labels.

### CODE REVIEW

File: cube/generic_networks/att_model.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
+import dynet as dy
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: Please add a License Header based on Apache V2. You can use as an example, one of the other files in NLP-Cube.

File: cube/generic_networks/att_model.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	 batch_size=2):
+
+		self.embedding_size = embedding_size
+		self.lstm_num_of_layers = lstm_num_of_layers
+		self.num_of_classes = num_of_classes
+		self.batch_size = batch_size
+		self.state_size = state_size
+		self.vocab_size = vocab_size + 1
+
+	def fit(self, X_train=None, y_train=None, X_test=None, y_test=None, epochs=None):
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: Please split this method and use a batch learning controlled from the parameters passed in main.py, namely --batch-size.

Also add a separate trainer for this, and display useful information such as:
- percentage during training (multiples of 5%)
- loss on the training set (after each epoch)
- epoch duration (execution time=...)
- a relevant metric on the development set after evaluation

Please store your models as follows:
 - use .last to indicate a checkpoint of the model after every epoch
 - use .best<Metric> to indicate a checkpoint for the best performing model on the development set.

I'm pasting here an excerpt  for you to use as an example:

```text
Starting training for parser
==PARAMETERS==
TRAIN FILE: ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-train.conllu
DEV FILE: ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-dev.conllu
EMBEDDINGS FILE: ../corpus/embeddings/wiki.dummy.vec
STOPPING CONDITION: 20
OUTPUT BASE: ../models/UD_English_SML/parser
AUX SOFTMAX WEIGHT: 0.2
CONFIG FILE: None
==============

Reading ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-train.conllu... found 12543 sequences
Reading ../corpus/ud-treebanks-v2.2/UD_English-EWT/en_ewt-ud-dev.conllu... found 2002 sequences
Computing encoding maps... done
Unique words: 17966
Unique chars: 75
Unique labels: 50
Unique UPOS: 17
Unique XPOS: 50
Holistic word count: 2954
200 303
Orthogonal pretrainer loss: 1.08e-26
200 303
Orthogonal pretrainer loss: 8.78e-27
300 400
Orthogonal pretrainer loss: 3.85e-23
300 400
Orthogonal pretrainer loss: 3.57e-23
300 900
Orthogonal pretrainer loss: 1.89e-30
300 900
Orthogonal pretrainer loss: 1.90e-30
200 800
Orthogonal pretrainer loss: 1.01e-30
200 800
Orthogonal pretrainer loss: 1.21e-30
200 600
Orthogonal pretrainer loss: 9.44e-31
200 600
Orthogonal pretrainer loss: 9.27e-31
200 600
Orthogonal pretrainer loss: 1.01e-30
200 600
Orthogonal pretrainer loss: 1.02e-30
Storing encodings in ../models/UD_English_SML/parser.encodings
Storing config in ../models/UD_English_SML/parser.conf
Starting epoch 1
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=89.07018480937295 execution_time=1129.7310800552368
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.5248131064100525 LAS=0.3065850167011293 UPOS=0.648043581994592 XPOS=0.530658501670113 ATTRS=0.58827739780$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 2
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=48.41013286851823 execution_time=1141.804716348648
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.6983060283123906 LAS=0.5644186416414825 UPOS=0.8104024176872913 XPOS=0.7141323365675203 ATTRS=0.749045649$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 3
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=34.46338212515743 execution_time=1136.4327113628387
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.7422061396532528 LAS=0.6536106250994115 UPOS=0.8678622554477493 XPOS=0.8225703833306823 ATTRS=0.843009384$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 4
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=28.836603429417647 execution_time=1133.9607968330383
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.7812549705742007 LAS=0.7145697470971847 UPOS=0.8923970097025609 XPOS=0.8686575473198664 ATTRS=0.889852075$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 5
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=25.650525833710162 execution_time=1138.0560524463654
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.8002226817241928 LAS=0.7442738985207571 UPOS=0.90726896771115 XPOS=0.8896134881501511 ATTRS=0.90929696198$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
Starting epoch 6
        shuffling training data... done
        training... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg_loss=23.388270108882086 execution_time=1138.4073255062103
        evaluating on devset... 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 UAS=0.8153729918880229 LAS=0.7618896134881501 UPOS=0.9118021313822173 XPOS=0.8984014633370447 ATTRS=0.915460473$
        Storing ../models/UD_English_SML/parser.bestUAS
        Storing ../models/UD_English_SML/parser.bestLAS
        Storing ../models/UD_English_SML/parser.last
```

File: cube/generic_networks/att_model.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		self.attention_w1 = self.model.add_parameters((self.state_size, self.state_size))
+		self.attention_w2 = self.model.add_parameters((self.state_size, self.state_size))
+		self.attention_v = self.model.add_parameters((1, self.state_size))
+
+		# Predictive allignment params
+		self.w_p = self.model.add_parameters((self.state_size, self.state_size))
+		self.v_p = self.model.add_parameters((1, self.state_size))
+
+		# LSTM/GRU and last layer projection matrix
+		self.lstm = dy.GRUBuilder(self.lstm_num_of_layers,
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: This formatting seems a little strange.

File: cube/generic_networks/att_model.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		for x in batch:
+			x = [self.vocab_size - 1] * (max_len - len(x)) + x
+			padded_batch.append(x)
+
+		return padded_batch
+
+	"""
+		Prepare data for training and testing.
+	"""
+	def _prepare_data(self, X_train, y_train, X_test, y_test):
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: Prepare your data in the Trainer

File: cube/generic_networks/att_model.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		for i in range(int(ceil(len(X) / self.batch_size))):
+			batched_X.append(data[0][i * self.batch_size : (i + 1) * self.batch_size])
+			batched_y.append(data[1][i * self.batch_size : (i + 1) * self.batch_size])
+
+		return batched_X, batched_y
+
+	"""
+		Pad sequences for efficient batching.
+	"""
+	def _pad_batch(self, batch):
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: Do you require padding for DyNet? You might get better results if you use variable sized sequences and you rely on the `--autobatch` parameter

File: cube/generic_networks/att_model.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		Predictive allignments based on importance of words to each class. Here W_p and v_p are
+		learnable parameters that get adjusted during training.
+	"""
+	def _predictive_allignment(self, sent_len, input_vector):
+		w_p = dy.parameter(self.w_p)
+		v_p = dy.parameter(self.v_p)
+
+		return sent_len * dy.logistic(v_p * dy.tanh(w_p * input_vector))
+
+	def _get_allignments(self, input_vectors):
```
Comment (tiberiu44) [inline cube/generic_networks/att_model.py]: Do you need to back propagate through this computation? If so, you should use DyNET primitives instead of numpy.

**Comment by tiberiu44** [comment]
I'm going to close this, due to inactivity.
